pkg/controller/cache/redis_test.go

```diff
@@ -184,713 +160,528 @@ func redisResource(rm ...redisResourceModifier) *v1alpha3.Redis {
 	return r
 }
 
-// Test that our Reconciler implementation satisfies the Reconciler interface.
-var _ reconcile.Reconciler = &Reconciler{}
+var _ resource.ExternalClient = &external{}
+var _ resource.ExternalConnecter = &connector{}
 
-func TestCreate(t *testing.T) {
-	cases := []struct {
-		name        string
-		csdk        createsyncdeletekeyer
-		r           *v1alpha3.Redis
-		want        *v1alpha3.Redis
-		wantRequeue bool
+func TestConnect(t *testing.T) {
+	type args struct {
+		cr          *v1alpha3.Redis
+		newClientFn func(ctx context.Context, credentials []byte) (redisapi.ClientAPI, error)
+		kube        client.Client
+	}
+	type want struct {
+		err error
+	}
+
+	cases := map[string]struct {
```

Comment:
I've been playing with a convention if including a `reason string` in each case that is printed when a test fails. The idea is that it makes it more obvious what the test is trying to do to future readers of the test code. So for example the `Successful` test might have `reason: "Connect should not return an error when we can successfully use the Provider and its Secret to produce a new client"`. Do you think it would be worth using this pattern for these tests?

Reply:
I feel like future code reader's effort to understand what the test does by looking at the case's name is less than the writer's effort to write a detailed reason section for each test. One might think writer does it only once so it's fine; but they'd do it for each and every case and reader doesn't go through all of them but only the ones they need, probably the ones that failed. 

Another note: getting people write tests is already hard, let's just not make it any harder.

Reply:
Hrm, I don't agree with this philosophy. I think it's generally better to optimise for code readers rather than writers.

Typically many more people will read and need to understand code during its lifetime than people will read it. For our tests we follow the Go philosophy of using (apart from `cmp`) just Go to write tests. I think this is the right choice because it means anyone who can understand Go can figure out our tests, but I do think it makes the _intent_ of each test a little harder to understand compared to testing DSLs that read like `Assert(Foo).Returns(NoError)` or similar. I think this has resulted in some cases us trying to encode our expectations in really long and complicated test names.

That said, this need not block this PR. I am happy to continue to try to sell this idea elsewhere. ;)